---
title: "Sample relationship verification using 450k DNA Methylation data: beta-values"
shorttitle: "Verifying sample relationships using beta-values"
package: "omicsPrint"
author: 
- name: "Maarten van Iterson"
  affiliation: 
  - "Department of Medical Statistics and Bioinformatics, Section Moleculair Epidemiology, Leiden University Medical Center"
  email: mviterson@gmail.com
date: "`r Sys.Date()`"
output: 
  BiocStyle::html_document2:
    toc: true
    toc_float:
      collapsed: false
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
bibliography: omicsPrint.bib
---

Here we will show how you could verify sample relationships on
publicly available DNA methylation data, often a beta-value matrix
from the Gene Expression Omnibus.

Here we use the beta-value matrix extracted from GEO (GSE52980). The
data was generated to perform an epigenome analysis of human epidermal
and dermal samples with aging and sun exposure[@Vandiver2015]. Since,
multiple samples from the same individual have been used we should be
able to verify this using genotypes extracted from the beta-values.


```{r setup, include=FALSE}
knitr::opts_chunk$set(eval=FALSE)
suppressPackageStartupMessages({
    library(omicsPrint)
    library(Biobase)
    library(GEOquery)
    library(DNAmArray) #see https://github.com/molepi/DNAmArray     
    library(minfi)
})
```

First we extract the data from GEO and select the platform on which
the DNA methylation was measured (GPL13534).

```{r, downloaddata}
library(Biobase)
library(GEOquery)
# load series and platform data from GEO
gset <- getGEO("GSE52980", GSEMatrix = TRUE, getGPL=FALSE)
idx <- grep("GPL13534", attr(gset, "names"))
betas <- exprs(gset[[idx]])
dim(betas)
```

We need to preprocess the metadata a little to obtain a clear picture
of which samples belong to which individuals. Furthermore, some
additional (carcinoma) samples have been removed.

```{r, extractphenotypes}
pheno <- pData(gset[[idx]])[, c("title", "characteristics_ch1", "characteristics_ch1.1",  "characteristics_ch1.2", "characteristics_ch1.3")]
colnames(pheno) <- c("title", "sex", "age", "exposure", "tissue")
pheno <- apply(pheno, 2, function(x) gsub("^.*: ", "", x))
pheno <- as.data.frame(pheno)
pheno <- subset(pheno, grepl("ermis", tissue)) ##ignore carcinoma samples
tbl <- table(pheno$age)
pheno <- subset(pheno, age %in% names(tbl[tbl == 4])) ##for ease of mapping to same individual
pheno <- pheno[order(pheno$age),]
nrow(pheno)/4
pheno$sample_id <- rep(1:15, each=4)
head(pheno)
betas <- betas[, colnames(betas) %in% rownames(pheno)]
mid <- match(colnames(betas), rownames(pheno))
colnames(betas) <- pheno$sample_id[mid]
dim(betas)
```

Now we can extract the genotypes. 

If we would have had the raw idats we could extract the 65
SNPs. However, experiences show that the 65 SNPs available on the
array are not enough to perform sample verification with high
confidence. Fortunately, several probes on the array contain SNPs
occuring frequently in different populations[@Chen2013; @Zhou2016]. We
have extended the results of Zhou *et al.* using information from the
dutch population [GoNL](http://www.nlgenome.nl/).

The paper reported that the individuals are from cauciasian origin so
we use a european set of SNPs; the GoNL set.

```{r, extractgenotypes}
library(DNAmArray)
data(hg19.GoNLsnps) ##From DNAmArray
cgs <- hg19.GoNLsnps$probe[hg19.GoNLsnps$distance_c <= 0]
dnamCalls <- beta2genotype(betas[rownames(betas) %in% cgs, ])
dnamCalls[1:5, 1:5]
dim(dnamCalls)
data <- alleleSharing(dnamCalls)
head(data)
mismatches <- inferRelations(data)
```

The related and unrelated pairs are separated very well and no
mismatches are detected.

# SessionInfo #

```{r, sessioninfo}
sessionInfo()
```

# Reference #
