---
title: "Verifying sample relationships using genotypes or inferred genotypes"
shorttitle: "Verifying sample relationships"
package: "omicsPrint"
author: 
- name: "Maarten van Iterson"
  affiliation: 
  - "Department of Medical Statistics and Bioinformatics, Section Moleculair Epidemiology, Leiden University Medical Center"
  email: mviterson@gmail.com
- name: "Davy Cats"
  affiliation: 
  - "Department of Medical Statistics and Bioinformatics, Section Moleculair Epidemiology, Leiden University Medical Center"
  email: davycats.dc@gmail.com
date: "`r Sys.Date()`"
abstract: >
  Analysis of multiple omics datatypes from the same individuals is
  becoming increasingly common. For example, several data repositories
  contain genetic, transcriptomic and epigenetic (DNA methylation)
  measurements on the same individuals, e.g. TCGA, Geuvadis,
  BBMRI/BIOS, etc. We have developed a tool that can verify the sample
  relationships between and across omics types. A small number of
  misspecified samples can destroy analyses for example assuming an
  analysis on unrelated individuals whereas parent offspring relations
  exist.
output: 
  BiocStyle::html_document2:
    toc: true
    toc_float:
      collapsed: false
vignette: >
  %\VignetteIndexEntry{Verifying sample relationships using genotypes or inferred genotypes}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
bibliography: omicsPrint.bib
---


```{r setup, include=FALSE}
set.seed(22062017)
library(omicsPrint)
library(GEOquery)
library(SummarizedExperiment)
```

```{r, downloaddata}	
library(GEOquery)
library(SummarizedExperiment)
##gset <- getGEO("GSE102177", GSEMatrix=TRUE, AnnotGPL=FALSE)
##this is more stable somehow?
file <- tempfile()
download.file("ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE102nnn/GSE102177/matrix/GSE102177_series_matrix.txt.gz", file)
gse <- getGEO(filename=file)
se <- makeSummarizedExperimentFromExpressionSet(gse)
se
```

```{r, omicsprint}
r <- expand.grid(idx=colnames(se), idy=colnames(se))
r$Xfam <- colData(se)[r$idx, "characteristics_ch1.3"]
r$Yfam <- colData(se)[r$idy, "characteristics_ch1.3"]

fun <- function(x){
    if (x["idx"] == x["idy"]){
        return("identical")
    } else if (x["Xfam"] == x["Yfam"]){
        return("sibling")
    } else {
        return("unrelated")
    }
}

r$relation_type <- apply(r, 1, fun)
head(r)

gt <- beta2genotype(se, assayName = "exprs")
data <- alleleSharing(gt, relations = r)
mismatches <- inferRelations(data)
mismatches
```

## Inspect Hardy-Weinberg filtering
```{r, hardyweinberg}
alpha <- seq(0, 1, by = .01)
nsnps <- numeric(length(alpha))
for(i in 1:length(alpha)) {
    x <- omicsPrint:::.pruning(gt, callRate=0.95, coverageRate=2/3, alpha = alpha[i], verbose=TRUE)
    nsnps[i] <- nrow(x)
}

nrow(gt)

par(mar=c(5,4,6,2))
plot(alpha/1000, 1002-nsnps, type = "b",
     log = "x", panel.first=grid(col=1, equilogs=FALSE),
     xlab = "Significant threshold (Bonferonni corrected)",
     ylab = "Number of SNPs",
     main = "Number of SNPs violating Hardy-Weinberg principle\n(at different significance levels)")
at <- axTicks(1)
axis(3, at, labels=at*1000)

```

```{r, snpsizes}	
set.seed(12345)
sizes <- c(5, 10, 25, 50, 100, 250, 500, 1000)

nrep <- 25
miss <- matrix(0, length(sizes), nrep)
cntr = 0
for(i in 1:length(sizes)){
    for(j in 1:nrep)     {
        cntr = cntr +1
        indices <- sample(1:nrow(gt), sizes[i])
        data <- alleleSharing(gt[indices,], relations = r)
        predictions <- inferRelations(data, plot.it=FALSE)
        miss[i, j] <- nrow(predictions) - sum(diag(table(predictions$predicted, predictions$relation)))
    }
}


rownames(miss) <- sizes
boxplot(t(miss),
        ylab="Number of missclassified relations (out of 666)", xlab="Number of SNPs")
```

	
# SessionInfo #

```{r, sessioninfo}
sessionInfo()
```
