---
title: "omicsPrint: detection of data linkage errors in multiple omics studies"
shorttitle: "Verifying sample relationships"
package: "omicsPrint"
author: 
- name: "Maarten van Iterson"
  affiliation: 
  - &id "Department of Medical Statistics and Bioinformatics, Section Moleculair Epidemiology, Leiden University Medical Center,
  Leiden, The Netherlands"  
  email: mviterson@gmail.com
- name: "Davy Cats"
  affiliation: *id  
  email: davycats.dc@gmail.com
- name: "Paul Hop"
  affiliation: *id    
  email: pjhop@gmail.com  
- name: "Bas T. Heijmans"
  affiliation: *id    
  email: b.t.heijmans@lumc.nl
date: "`r Sys.Date()`"
abstract: >
  The Aanalysis of multiple omics datatypes from the same
  individuals is becoming increasingly common. For example,
  several data repositories contain genetic, transcriptomic
  and epigenetic (DNA methylation) measurements on the same
  individuals, e.g., TCGA, Geuvadis, BBMRI/BIOS, etc. However,
  errors in the data linkage are almost inevitable in such
  complex projects (e.g. due to sample mix-ups) and unknown
  family relationships may be present. If not corrected, such
  errors may introduce false positive or false negative
  findings. To streamline necessary quality control, we have
  developed a tool to reliably verify the sample relationships
  between and across omics data types using genotype data that
  is directly measured or inferred from for example RNA-seq or
  DNA methylation array data.    	  
output: 
  BiocStyle::html_document:
    toc: true
    toc_float:
      collapsed: false
vignette: >
  %\VignetteIndexEntry{omicsPrint: detection of data linkage errors in multiple omics studies}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
bibliography: omicsPrint.bib
---


## Retrieving sample information
```{r}
library(GEOquery)
##gset <- getGEO("GSE100227", GSEMatrix=TRUE, AnnotGPL=FALSE)
##this is more stable somehow?
file <- tempfile()
download.file("ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE100nnn/GSE100227/matrix/GSE100227_series_matrix.txt.gz", file)
gse <- getGEO(filename = file)
gse
samples <- pData(gse)
rownames(samples) <- samples$description
```

## Retrieving beta values
Unfortunately the data retrieved from GEO above does not include the beta values.
These values instead as provided in a supplementary file.

```{r, warning=FALSE}
library(readr)

file2 <- tempfile()
file3 <- tempfile()
download.file("ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE100nnn/GSE100227/suppl/GSE100227_normalized_data.txt.gz", file2)
gunzip(file2, file3)

d <- read_tsv(file3)
d <- d[, ! grepl("Detection", colnames(d))] # remove useless columns
d <- as.data.frame(d)
rownames(d) <- d$ID_REF
d <- d[, !colnames(d) == "ID_REF"]
```

## Determine relationships
```{r}
r <- expand.grid(idx=rownames(samples), idy=rownames(samples))

r$Xfam <- substr(samples[r$idx, "characteristics_ch1.1"], 12, 100)
r$Yfam <- substr(samples[r$idy, "characteristics_ch1.1"], 12, 100)
r$Xrole <- substr(samples[r$idx, "characteristics_ch1.2"], 11, 100)
r$Yrole <- substr(samples[r$idy, "characteristics_ch1.2"], 11, 100)

fun <- function(x){
    if (x["idx"] == x["idy"]) {
        return("identical")
    } else if (x["Xfam"] != x["Yfam"]) {
        return("unrelated")
    } else if (x["Xrole"] == x["Yrole"]) {
        return(x["Xrole"])
    } else {
        return("Sister")
    }
    return("hello")
}

r$relation_type <- apply(r, 1, fun)
head(r)
```

## Filter probes
```{r}
library(omicsPrint)
library(GenomicRanges)
data(hm450.manifest.pop.GoNL)
cpgs <- names(hm450.manifest.pop.GoNL[
        mcols(hm450.manifest.pop.GoNL)$MASK.snp5.EUR])

d2 <- d[cpgs,]
```

## Check relationships
```{r}
gt <- beta2genotype(d2)
data <- alleleSharing(gt, relations = r, verbose = TRUE)
mismatches <- inferRelations(data)
```

# SessionInfo #

```{r, sessioninfo}
sessionInfo()
```

# Reference #
