---
title: "omicsPrint: detection of data linkage errors in multiple omics studies"
shorttitle: "Verifying sample relationships"
package: "omicsPrint"
author: 
- name: "Maarten van Iterson"
  affiliation: 
  - &id "Department of Medical Statistics and Bioinformatics, Section Moleculair Epidemiology, Leiden University Medical Center,
  Leiden, The Netherlands"  
  email: mviterson@gmail.com
- name: "Davy Cats"
  affiliation: *id  
  email: davycats.dc@gmail.com
- name: "Paul Hop"
  affiliation: *id    
  email: pjhop@gmail.com  
- name: "Bas T. Heijmans"
  affiliation: *id    
  email: b.t.heijmans@lumc.nl
date: "`r Sys.Date()`"
abstract: >
  The Aanalysis of multiple omics datatypes from the same
  individuals is becoming increasingly common. For example,
  several data repositories contain genetic, transcriptomic
  and epigenetic (DNA methylation) measurements on the same
  individuals, e.g., TCGA, Geuvadis, BBMRI/BIOS, etc. However,
  errors in the data linkage are almost inevitable in such
  complex projects (e.g. due to sample mix-ups) and unknown
  family relationships may be present. If not corrected, such
  errors may introduce false positive or false negative
  findings. To streamline necessary quality control, we have
  developed a tool to reliably verify the sample relationships
  between and across omics data types using genotype data that
  is directly measured or inferred from for example RNA-seq or
  DNA methylation array data.    	  
output: 
  html_document:
    toc: true
    toc_float:
      collapsed: false
vignette: >
  %\VignetteIndexEntry{omicsPrint: detection of data linkage errors in multiple omics studies}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
bibliography: omicsPrint.bib
---

# Sample verification using 450k DNA methylation from GEO #

Here we use the beta-value matrix extracted from GEO (GSE52980). The
data was generated to perform an epigenome analysis of human epidermal
and dermal samples with aging and sun exposure[@Vandiver2015]. Since,
multiple samples from the same individual have been used we should be
able to verify this using genotypes extracted from the beta-values.


```{r setup, include=FALSE}
suppressPackageStartupMessages({
    library(BiocStyle)
    library(omicsPrint)
    library(GEOquery)
})
```

# Extract beta-values from GEO #

First we extract the data from GEO and select the platform on which
the DNA methylation was measured (GPL13534).

```{r, downloaddata}
library(GEOquery)
##gse <- getGEO("GSE52980", GSEMatrix=TRUE, AnnotGPL=FALSE)
##this is more stable somehow?
file <- tempfile()
download.file("ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE52nnn/GSE52980/matrix/GSE52980-GPL13534_series_matrix.txt.gz", file)
gse <- getGEO(filename=file)

betas <- exprs(gse)
dim(betas)

```

# Preprocessing metadata #

We need to preprocess the metadata a little to obtain a clear picture
of which samples belong to which individuals. Furthermore, some
additional (carcinoma) samples have been removed.

```{r, extractphenotypes}
pheno <- pData(gse)[, c("title", "characteristics_ch1", 
    "characteristics_ch1.1",  "characteristics_ch1.2", "characteristics_ch1.3")]
colnames(pheno) <- c("title", "sex", "age", "exposure", "tissue")
pheno <- apply(pheno, 2, function(x) gsub("^.*: ", "", x))
pheno <- as.data.frame(pheno)
pheno <- subset(pheno, grepl("ermis", tissue)) ##ignore carcinoma samples
tbl <- table(pheno$age)
pheno <- subset(pheno, age %in% names(tbl[tbl == 4])) ##for ease of mapping to same individual
pheno <- pheno[order(pheno$age),]
nrow(pheno)/4
pheno$sample_id <- rep(1:15, each=4)
head(pheno)
betas <- betas[, colnames(betas) %in% rownames(pheno)]
mid <- match(colnames(betas), rownames(pheno))
colnames(betas) <- pheno$sample_id[mid]
dim(betas)
```

# Convert beta-values to genotypes #

If we would have had the raw idats we could extract the 65
SNPs. However, experiences show that the 65 SNPs available on the
array are not enough to perform sample verification with high
confidence. Fortunately, several probes on the array contain SNPs
occuring frequently in different populations[@Chen2013; @Zhou2016]. We
have extended the results of Zhou *et al.* using information from the
dutch population [GoNL](http://www.nlgenome.nl/).

The paper reported that the individuals are from cauciasian origin so
we use a european set of SNPs; the GoNL set.

```{r, extractgenotypes}
library(omicsPrint)
data(hg19.GoNLsnps)
cgs <- hg19.GoNLsnps$probe[hg19.GoNLsnps$distance_c <= 0]
dnamCalls <- beta2genotype(betas[rownames(betas) %in% cgs, ])
dnamCalls[1:5, 1:5]
dim(dnamCalls)
```

```{r, allelesharing}
data <- alleleSharing(dnamCalls)
head(data)
mismatches <- inferRelations(data)
```

The related and unrelated pairs are separated very well and no
mismatches are detected.

# SessionInfo #

```{r, sessioninfo}
sessionInfo()
```

# Reference #
